### PR TITLE
Adding BigQuery Storage direct write support

### PR DESCRIPTION
**Pull request by davidrabinowitz**

### CODE REVIEW

**Comment by davidrabinowitz** [comment]
Notice that DataSourceV1IndirectWriteIntegrationTest is WIP

File: spark-bigquery-connector-common/third_party/apache-spark/src/main/java/com/google/cloud/spark/bigquery/AvroSchemaConverter.java [hunk 1]
```diff
@@ -187,15 +194,33 @@ static Converter createConverterFor(DataType sparkType, Schema avroType) {
     if (sparkType instanceof DecimalType && avroType.getType() == Schema.Type.BYTES) {
       DecimalType decimalType = (DecimalType) sparkType;
       return (getter, ordinal) -> {
-        Decimal decimal = getter.getDecimal(ordinal, decimalType.precision(), decimalType.scale());
+        BigDecimal bigDecimal = null;
+        if (getter instanceof UnsafeRow || getter instanceof UnsafeArrayData) {
```
Comment (VinayLondhe14) [inline spark-bigquery-connector-common/third_party/apache-spark/src/main/java/com/google/cloud/spark/bigquery/AvroSchemaConverter.java:198]: Similar to the above question, were these changes required due to the direct write support?

File: spark-bigquery-connector-common/src/main/java/com/google/cloud/spark/bigquery/ProtobufUtils.java [hunk 2]
```diff
@@ -321,48 +320,70 @@ private static Object convertSparkValueToProtoRowValue(
...
+        || sparkType instanceof LongType) {
       return ((Number) sparkValue).longValue();
+    }
+    if (sparkType instanceof TimestampType) {
+      return SparkBigQueryUtil.sparkTimestampToBigQuery(sparkValue);
     } // TODO: CalendarInterval
 
     if (sparkType instanceof DateType) {
-      return ((Number) sparkValue).intValue();
+      return SparkBigQueryUtil.sparkDateToBigQuery(sparkValue);
```
[the remark this answers is not in the document]
Comment (davidrabinowitz) [reply spark-bigquery-connector-common/src/main/java/com/google/cloud/spark/bigquery/ProtobufUtils.java:374]: We are converting them to the BigQuery format. The problem is that based on the DataSource version, we are getting different types (Long vs java.sql.Timestamp) and need to provide the same long that BQ expects. The util method takes case of it

File: bigquery-connector-common/src/main/java/com/google/cloud/bigquery/connector/common/BigQueryClient.java [hunk 3]
```diff
@@ -221,10 +226,9 @@ String sqlFromFormat(String queryFormat, TableId destinationTableId, TableId tem
    */
   public String createTablePathForBigQueryStorage(TableId tableId) {
     Preconditions.checkNotNull(tableId, "tableId cannot be null");
-    Preconditions.checkNotNull(tableId.getProject(), "tableId.project cannot be null");
+    String project = tableId.getProject() != null ? tableId.getProject() : getProjectId();
```
[the remark this answers is not in the document]
Comment (davidrabinowitz) [reply bigquery-connector-common/src/main/java/com/google/cloud/bigquery/connector/common/BigQueryClient.java:229]: We need the full path for the createWriteStream method. We used to have it by creating the table and then taking its full tableId, but that caused an issue with the ErrorIfExists implementation (now the check, done in another place is positive). To solve it, we do what the BigQuery client does on Table ID with no project - take the BigQuery client own project ID.  This gives us the same behavior but allows us to defer the table creation to the last minute.
